src/test/java/com/microsoft/azure/spark/tools/http/HttpObservableScenario.java

```diff
@@ -0,0 +1,204 @@
+    }
+
+    @Given("^add a cookie '(.*):(.*)' into cookie store")
+    public void mockCookieStore(String name, String value) {
+        BasicClientCookie cookie = new BasicClientCookie(name, value);
+        cookie.setExpiryDate(DateTime.now().plus(Period.days(1)).toDate());
+        cookie.setDomain("localhost");
+        cookieStore.addCookie(cookie);
+
+        this.httpMock.setCookieStore(this.cookieStore);
+    }
+
+    @Then("^send and check Http '(.+)' request to '(.*)' should contain headers$")
+    public void checkRequestHeaderAndParameters(String method,
+                                                String path,
+                                                Map<String, String> expectHeaders) {
+        this.httpMock.request(this.httpRequest, null, null, null ).subscribe();
+
+        RequestPatternBuilder builder = RequestPatternBuilder.newRequestPattern(RequestMethod.fromString(method), urlPathEqualTo(path));
+
+        expectHeaders.forEach((k, v) -> builder.withHeader(k, equalTo(v)));
+
+        verify(builder);
+    }
+
+    @Then("^send and check Https '(.+)' request to '(.*)' should contain headers$")
+    public void checkHttpsRequestHeaderAndParameters(String method,
+                                                String path,
+                                                Map<String, String> expectHeaders) {
+        this.httpMock.request(this.httpRequest, null, null, null ).subscribe();
+
+        RequestPatternBuilder builder = RequestPatternBuilder.newRequestPattern(RequestMethod.fromString(method), urlPathEqualTo(path));
+
+        expectHeaders.forEach((k, v) -> builder.withHeader(k, equalTo(v)));
+
+        verify(builder);
+    }
+
+    @Then("send and check Https {string} request to {string} should throw CertificateException {string}")
+    public void checkHttpsRequestValidatorException(String method,
+                                                    String path,
+                                                    String expectErrorMessage) {
+        AtomicReference<Throwable> caught = new AtomicReference<>();
+
+        this.httpMock.request(this.httpRequest, null, null, null)
+                .toBlocking()
+                .subscribe(
+                        data -> {},
+                        caught::set
+                );
+
+
+        assertTrue(
+                "The exception caught was not a CertificateException: " + ExceptionUtils.getMessage(caught.get()),
+                caught.get().getCause() instanceof CertificateException);
+        assertEquals(expectErrorMessage, caught.get().getCause().getMessage());
+    }
+
+    @Then("send and check Https {string} request to {string} should throw SSLException {string}")
+    public void checkHttpsRequestSSLException(String method,
+                                                    String path,
+                                                    String expectErrorMessage) {
+        AtomicReference<Throwable> caught = new AtomicReference<>();
```

Comment:
Just in curious, why do we need to make this filed `AtomicReference`

Reply:
Since it's wrote in an lambda function. 😄